### PR TITLE
[RELAY][FRONTEND] Tensorflow frontend.

### PR DESCRIPTION
**Pull request by srkreddy1238**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

### CODE REVIEW

File: tests/python/frontend/tensorflow/test_forward.py [hunk 1]
```diff
@@ -0,0 +1,1111 @@
+# pylint: disable=import-self, invalid-name, unused-argument
+"""
+Tensorflow testcases
+====================
+This article is a test script to test tensorflow operator with NNVM.
```
Comment (jroesch) [inline tests/python/frontend/tensorflow/test_forward.py]: Should we replace occurrences to NNVM with Relay?

**Comment by srkreddy1238** [comment]
@jroesch @Huyuwei & @tqchen  welcome to review.
Except LSTM now it's equivalent to NNVM frontend.

LSTM is suffering with some strange errors, that too different errors for python2 (LLVM/FCmp) and python3(schedule_injective). I am working on it....

**Comment by srkreddy1238** [comment]
@jroesch  thanks for the review. 
I enhanced with docstrings cleanup and moved the testing.tf to relay. You may have another look.
#2382 is a dependency for this PR. Now except LSTM everything work good.

I came across some strange error for LSTM, yet to debug

Does the below log ring any bell ??

```
Traceback (most recent call last):
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/relay/backend/compile_engine.py", line 78, in lower
    return _backend._CompileEngineLower(self, key)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 185, in __call__
    ctypes.byref(ret_val), ctypes.byref(ret_tcode)))
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/base.py", line 72, in check_call
    raise TVMError(py_str(_LIB.TVMGetLastError()))
tvm._ffi.base.TVMError: TVMCall CFunc Error:
Traceback (most recent call last):
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 55, in cfun
    rv = local_pyfunc(*pyargs)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/relay/op/op.py", line 186, in schedule_injective
    return topi.generic.schedule_injective(outputs)
  File "<decorator-gen-62>", line 2, in schedule_injective
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/target.py", line 273, in dispatch_func
    return generic_func_node(*args)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/target.py", line 135, in __call__
    return _api_internal._GenericFuncCallFunc(self, *args)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 185, in __call__
    ctypes.byref(ret_val), ctypes.byref(ret_tcode)))
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/base.py", line 72, in check_call
    raise TVMError(py_str(_LIB.TVMGetLastError()))
tvm._ffi.base.TVMError: TVMCall CFunc Error:
Traceback (most recent call last):
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 55, in cfun
    rv = local_pyfunc(*pyargs)
  File "/home/srk/.local/lib/python3.6/site-packages/topi-0.5.dev0-py3.6.egg/topi/x86/injective.py", line 26, in schedule_injective
    if len(s[x].op.axis) >= 5:
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/node.py", line 59, in __getattr__
    "'%s' object has no attribute '%s'" % (str(type(self)), name))
AttributeError: '<class 'tvm.tensor.PlaceholderOp'>' object has no attribute 'axis'

```

**Comment by jroesch** [comment]
It looks like a bug in the schedule, maybe fusion is introducing a pattern the master schedule is not written for? Do you know what operator this is failing for? we should probably modify the compile engine to dump more debug information when an exception happens while scheduling.

**Comment by srkreddy1238** [comment]
@jroesch 
Above situation is caused due to Concatenate operator with one Tensor in list. For now I handled it in front end.
We could decide to make Concatenate behave nice with this situation or leave it.

**Comment by srkreddy1238** [comment]
@jroesch 
Concatenate issue persists still. It's reproducible when place holders are passed to concatenate.

```
Traceback (most recent call last):
  File "./tests/python/frontend/tensorflow/test_forward.py", line 1059, in <module>
    test_forward_ptb()
  File "./tests/python/frontend/tensorflow/test_forward.py", line 871, in test_forward_ptb
    params, m = _get_tvm_graph_module(graph_def)
  File "./tests/python/frontend/tensorflow/test_forward.py", line 817, in _get_tvm_graph_module
    graph, lib, params = relay.build(sym, target, params=params)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/relay/build_module.py", line 241, in build
    graph_json, lowered_funcs, params = graph_gen.codegen(func)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/relay/backend/graph_runtime_codegen.py", line 349, in codegen
    self.heads = self.visit(func.body)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/relay/expr_functor.py", line 27, in visit
    res = self.visit_call(expr)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/relay/backend/graph_runtime_codegen.py", line 235, in visit_call
    cached_func = self.compile_engine.lower(func, self.target)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/relay/backend/compile_engine.py", line 86, in lower
    raise RuntimeError(msg)
RuntimeError: Traceback (most recent call last):
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/relay/backend/compile_engine.py", line 78, in lower
    return _backend._CompileEngineLower(self, key)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 185, in __call__
    ctypes.byref(ret_val), ctypes.byref(ret_tcode)))
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/base.py", line 72, in check_call
    raise TVMError(py_str(_LIB.TVMGetLastError()))
tvm._ffi.base.TVMError: TVMCall CFunc Error:
Traceback (most recent call last):
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 55, in cfun
    rv = local_pyfunc(*pyargs)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/relay/op/op.py", line 186, in schedule_injective
    return topi.generic.schedule_injective(outputs)
  File "<decorator-gen-62>", line 2, in schedule_injective
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/target.py", line 273, in dispatch_func
    return generic_func_node(*args)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/target.py", line 135, in __call__
    return _api_internal._GenericFuncCallFunc(self, *args)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 185, in __call__
    ctypes.byref(ret_val), ctypes.byref(ret_tcode)))
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/base.py", line 72, in check_call
    raise TVMError(py_str(_LIB.TVMGetLastError()))
tvm._ffi.base.TVMError: TVMCall CFunc Error:
Traceback (most recent call last):
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 55, in cfun
    rv = local_pyfunc(*pyargs)
  File "/home/srk/.local/lib/python3.6/site-packages/topi-0.5.dev0-py3.6.egg/topi/x86/injective.py", line 26, in schedule_injective
    if len(s[x].op.axis) >= 5:
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/node.py", line 59, in __getattr__
    "'%s' object has no attribute '%s'" % (str(type(self)), name))
AttributeError: '<class 'tvm.tensor.PlaceholderOp'>' object has no attribute 'axis'


Error during compile func
--------------------------
fn (%p0: Tensor[(1, 10000), float32],
    %p1: Tensor[(1, 2, 1, 200), float32],
    %p2: Tensor[(1, 2, 1, 200), float32])
    -> Tuple[Tensor[(1, 10000), float32], Tensor[(2, 2, 1, 200), float32]] {
  %0 = (%p1, %p2)
  %1 = concatenate(%0) # ty=Tensor[(2, 2, 1, 200), float32]
  %2 = (%p0, %1)
  %2
}
--------------------------
```

**Comment by srkreddy1238** [comment]
@jroesch  #2412 fixes the above problem.

But LLVM has below issue with fuse. Disabling fuse works fine though.

```
Traceback (most recent call last):
  File "tests/python/frontend/tensorflow/test_forward.py", line 1060, in <module>
    _test_forward_ptb()
  File "tests/python/frontend/tensorflow/test_forward.py", line 872, in _test_forward_ptb
    params, m = _get_tvm_graph_module(graph_def)
  File "tests/python/frontend/tensorflow/test_forward.py", line 818, in _get_tvm_graph_module
    graph, lib, params = relay.build(sym, target, params=params)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/relay/build_module.py", line 242, in build
    mod = _tvm_build_module(lowered_funcs, target=target, target_host=target_host)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/build_module.py", line 592, in build
    mhost = codegen.build_module(fhost_all, str(target_host))
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/codegen.py", line 20, in build_module
    return _Build(lowered_func, target)
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 185, in __call__
    ctypes.byref(ret_val), ctypes.byref(ret_tcode)))
  File "/home/srk/.local/lib/python3.6/site-packages/tvm-0.5.dev0-py3.6-linux-x86_64.egg/tvm/_ffi/base.py", line 72, in check_call
    raise TVMError(py_str(_LIB.TVMGetLastError()))
tvm._ffi.base.TVMError: [12:16:42] /home/srk/work/DMLC/tvm/src/codegen/llvm/llvm_module.cc:173: LLVM module verification failed with the following errors: 
Invalid operand types for FCmp instruction
  %293 = fcmp oeq i8* %34, %292
```

**Comment by srkreddy1238** [comment]
All test cases pass now. Above issue specific to PTB model is unrelated to frontend. 
I disabled optimisation only for that test case for now.
@Huyuwei @zhreshold welcome to review.

**Comment by tqchen** [comment]
@jroesch @Huyuwei @kazum please take a look and https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

File: python/tvm/relay/frontend/tensorflow.py [hunk 2]
```diff
@@ -0,0 +1,1539 @@
...
+                op_name="reshape",
+                extras={'newshape':tuple(shape_arg.asnumpy())},
+                ignores=['Tshape'])(inputs, attr)
+        except KeyError:
+            # Shape operator is already pruned, hence
+            # try to infer shape by precompute prune if possible.
+            if all(in_node in params for in_node in inputs[1].list_input_names()):
+                graph = _graph.create(_op.Group(inputs[1]))
+                params_pre = {k: params[k] for k in inputs[1].list_input_names()}
+                params_new = build_module._run_graph(graph, params_pre)
```
Comment (kazum) [inline python/tvm/relay/frontend/tensorflow.py]: `_graph` and `build_module` are not defined.

File: python/tvm/relay/frontend/tensorflow.py [hunk 3]
```diff
@@ -0,0 +1,1539 @@
...
+                    op = [_expr.var(node.name,
+                                    shape=self._params[node.name].shape,
+                                    dtype=self._params[node.name].dtype)]
+
+                elif isinstance(op, (_expr.TupleWrapper, tuple, list)):
+                    pass
+                elif isinstance(op, _expr.Expr):
+                    op = [op]
+                else:
+                    raise RuntimeError("unexpected type %s" % type(res))
```
Comment (kazum) [inline python/tvm/relay/frontend/tensorflow.py]:
```suggestion
                    raise RuntimeError("unexpected type %s" % type(op))
```

File: tutorials/frontend/from_tensorflow.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+# os and numpy
+import numpy as np
+import os.path
+
+# Tensorflow imports
+import tensorflow as tf
+from tensorflow.core.framework import graph_pb2
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import tensor_util
```
Comment (kazum) [inline tutorials/frontend/from_tensorflow.py]: No need to import `graph_pb2`, `dtypes`, and `tensor_util`.  These are not used.

**Comment by tqchen** [comment]
@srkreddy1238 please update, @kazum @jroesch please take another look and https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

File: python/tvm/relay/frontend/tensorflow.py [hunk 5]
```diff
@@ -0,0 +1,1540 @@
...
+            return AttrCvt(
+                op_name="reshape",
+                extras={'newshape':tuple(shape_arg.asnumpy())},
+                ignores=['Tshape'])(inputs, attr)
+        except KeyError:
+            # Shape operator is already pruned, hence
+            # try to infer shape by precompute prune if possible.
+            if all(in_node in params for in_node in inputs[1].list_input_names()):
+                func = _expr.Function(ir_pass.free_vars(inputs[1]), inputs[1])
+                with relay.build_config(opt_level=0):
```
Comment (kazum) [inline python/tvm/relay/frontend/tensorflow.py]:
```suggestion
                with tvm.relay.build_config(opt_level=0):
```

File: python/tvm/relay/frontend/tensorflow.py [hunk 6]
```diff
@@ -0,0 +1,1540 @@
...
+                    graph, lib, params = relay.build(func, target="llvm", params=params)
```
Comment (kazum) [inline python/tvm/relay/frontend/tensorflow.py]:
```suggestion
                    graph, lib, params = tvm.relay.build(func, target="llvm", params=params)
```

File: tests/python/frontend/tensorflow/test_forward.py [hunk 7]
```diff
@@ -0,0 +1,1113 @@
...
+            for i in range(len(tf_output)):
+                tvm.testing.assert_allclose(tf_output[i], tvm_output[i], atol=1e-5, rtol=1e-5)
+
+        sess.close()
+
+def is_gpu_available():
+    from tensorflow.python.client import device_lib
+    local_device_protos = device_lib.list_local_devices()
+    gpu_list = [x.name for x in local_device_protos if x.device_type == 'GPU']
+    if len(gpu_list) < 0:
```
Comment (kazum) [inline tests/python/frontend/tensorflow/test_forward.py]:
```suggestion
    if len(gpu_list) > 0:
```

File: tests/python/frontend/tensorflow/test_forward.py [hunk 8]
```diff
@@ -0,0 +1,1113 @@
...
+    with tf.Graph().as_default():
+        gen_array_ops._concat_v2(data, dim)
+
+        compare_tf_with_tvm(data, ['ConcatV2/values_0:0', 'ConcatV2/values_1:0'],
+                            'ConcatV2:0')
+
+def _test_forward_concat_v2():
+    t1 = np.array([])
+    t2 = np.array([])
+    test_concat_v2([t1, t2], 0)
```
Comment (kazum) [inline tests/python/frontend/tensorflow/test_forward.py]:
```suggestion
    _test_concat_v2([t1, t2], 0)
```

File: tests/python/frontend/tensorflow/test_forward.py [hunk 9]
```diff
@@ -0,0 +1,1113 @@
...
+
+        img_array = np.random.uniform(size=(1, 600, 600, 3)).astype("uint8")
+        img = Image.frombuffer('RGB', (600, 600), img_array.tostring(), 'raw', 'RGB', 0, 1)
+        temp = util.tempdir()
+        img_path = temp.relpath("tf-test.jpg")
+        img.save(img_path);
+
+        import os.path
+        if not tf.gfile.Exists(os.path.join(img_path)):
+            tf.logging.fatal('File does not exist %s', image)
```
Comment (kazum) [inline tests/python/frontend/tensorflow/test_forward.py]:
```suggestion
            tf.logging.fatal('File does not exist %s', img_path)
```

**Comment by srkreddy1238** [comment]
@ZihengJiang & @yzhliu request to consider this for 0.5 release if possible.

**Comment by tqchen** [comment]
@srkreddy1238 please look into http://ci.tvm.ai:8080/blue/organizations/jenkins/tvm/detail/PR-2216/32/pipeline

@kazum @jroesch please take another look

**Comment by kazum** [comment]
@srkreddy1238 My previous comments are not addressed in the latest version.

**Comment by srkreddy1238** [comment]
@yzhliu no problem :)

**Comment by yzhliu** [comment]
Thanks everyone for contribution and reviewing. Let's merge.

**Pull request merged**